compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -577,12 +578,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 // We have `&T`, check if what was expected was `T`. If so,
                 // we may want to suggest removing a `&`.
                 if sm.is_imported(expr.span) {
-                    if let Ok(src) = sm.span_to_snippet(sp) {
-                        if let Some(src) = src.strip_prefix('&') {
+                    // Go through the spans from which this span was expanded,
+                    // and find the one that's pointing inside `sp`.
+                    //
+                    // E.g. for `&format!("")`, where we want the span to the
+                    // `format!()` invocation instead of its expansion.
+                    if let Some(call_span) =
+                        iter::successors(Some(expr.span), |s| s.parent()).find(|&s| sp.contains(s))
```

Comment:
@estebank What do you think about adding as a function on `Span` (e.g. `span.find_parent_within(span)` or something)?

I have a feeling more diagnostics can/should use this, but I haven't written enough diagnostics yet to know if that's true.

Reply:
I agree with your assessment and we should have something like that (bikeshed notwithstanding).